contracts/examples/bonding-curve/src/utils/user_endpoints.rs

```diff
@@ -0,0 +1,264 @@
+
+		self.nonce_amount(&offered_token, nonce)
+			.update(|val| *val += sell_amount);
+
+		self.send().direct(
+			&caller,
+			&self.bonding_curve(&offered_token).get().payment_token,
+			0u64,
+			&calculated_price,
+			b"selling",
+		);
+		self.token_details(&offered_token)
+			.update(|details| details.add_nonce(nonce));
+
+		self.sell_token_event(&caller, &calculated_price);
+
+		Ok(())
+	}
+
+	#[payable("*")]
+	#[endpoint(buyToken)]
+	fn buy_token(
+		&self,
+		#[payment_amount] payment: Self::BigUint,
+		#[payment_token] offered_token: TokenIdentifier,
+		requested_amount: Self::BigUint,
+		requested_token: TokenIdentifier,
+		#[var_args] requested_nonce: OptionalArg<u64>,
+	) -> SCResult<()> {
+		let payment_token =
+			self.check_owned_return_payment_token(&requested_token, &requested_amount)?;
+		self.check_given_token(&payment_token, &offered_token)?;
+
+		let calculated_price = self
+			.bonding_curve(&requested_token)
+			.update(|bonding_curve| {
+				let price = self.compute_buy_price(
+					&bonding_curve.curve,
+					requested_amount.clone(),
+					bonding_curve.arguments.clone(),
+				);
+				let price_clone = price.clone()?;
+				require!(
+					price_clone <= payment,
+					"The payment provided is not enough for the transaction"
+				);
+				bonding_curve.payment_amount += price_clone;
+				bonding_curve.arguments.balance -= &requested_amount;
+
+				price
+			})?;
+
+		let caller = self.blockchain().get_caller();
+
+		match requested_nonce {
+			OptionalArg::Some(nonce) => {
+				self.send().direct(
+					&caller,
+					&requested_token,
+					nonce,
+					&requested_amount,
+					b"buying",
+				);
+				if self.nonce_amount(&requested_token, nonce).get() - requested_amount.clone() > 0 {
+					self.nonce_amount(&requested_token, nonce)
+						.update(|val| *val -= requested_amount.clone());
+				} else {
+					self.nonce_amount(&requested_token, nonce).clear();
+					self.token_details(&requested_token)
+						.update(|details| details.remove_nonce(nonce))?;
+				}
+			},
+			OptionalArg::None => {
+				self.send_bought_tokens(&caller, requested_token, requested_amount)?;
+			},
+		};
+
+		self.send().direct(
+			&caller,
+			&offered_token,
+			0u64,
+			&(&payment - &calculated_price),
+			b"rest",
+		);
+
+		self.buy_token_event(&caller, &calculated_price);
+		Ok(())
+	}
+
+	fn send_bought_tokens(
+		&self,
+		caller: &Address,
+		token: TokenIdentifier,
+		amount: Self::BigUint,
+	) -> SCResult<()> {
+		let mut nonces = self.token_details(&token).get().token_nonces;
+		let mut total_amount = amount;
+		loop {
+			let nonce = *nonces.first().ok_or("Requested nonce does not exist")?;
+			let available_amount = self.nonce_amount(&token, nonce).get();
+
+			let amount_to_send: Self::BigUint;
+			if available_amount <= total_amount {
+				amount_to_send = available_amount.clone();
+				total_amount -= amount_to_send.clone();
+				self.nonce_amount(&token, nonce).clear();
+				nonces.remove(0);
```

Comment:
I think there's a `pop_first` method or similar for Vec, so you can use that instead.

Reply:
nope